_posts/2019-01-04-turn-craft-into-a-headless-cms.md

```diff
@@ -0,0 +1,169 @@
+---
+title: 'Turn Craft 3 into a headless CMS'
+author: dimitri-steinel
+layout: post
+mainImage: "http://res.cloudinary.com/dsteinel/image/upload/c_scale,w_1012/v1536323225/Espi/blog/heroku-craft/craft-heroku-quickstart.jpg"
+mainImageAlt: "Yes this is a rocket"
+--- 
+
+
+### What is a headless CMS?
+The classic CMS approach is to couple the presentational layer with the CMS directly. With this old-fashioned monolithic approach of building a CMS, your developers will be bound to a specific template language and you lose a lot of flexibility. These kind of CMS’s are meant to output their content on only one channel: the Webbrowser.
+You may think that this is enough for now, but we will prove that you can do so much more with your content than just show it in one channel. We want to create an omnichannel experience.
+
+### Why would I want to have a headless CMS?
+**Performance, Security and Scalability.**
+<br />
+In the headless approach there is no presentational layer out-of-the-box and therefore developers are not bound to any templating or programming language to create one. This means, that a headless CMS can even have multiple heads and this gives us a lot of freedom, scalability and flexibility. Think of an App, Voice Interface, AR or any other kind of new technology which comes to the market. 
+<br />
+<br />
+A Static Site Generator like [GatsbyJS](https://www.gatsbyjs.org/) or [Nuxt.js](https://nuxtjs.org/) (there are [many others](https://www.staticgen.com/)) fetches all the content from your CMS via an API and generates old fashioned static sites. Static does not mean the site can not be interactive. It just means, that the site doesn't need to communicate with the CMS to fetch the content. If you are looking to improve the security of your site, a static site can live on an other server without any connection to the CMS. The connection happens during build time which means, our frontend will be blazing fast!
+
+### CraftQL
+Our static site generator of choice is GatsbyJS. With the plugin [CraftQL](https://github.com/markhuot/craftql) we can connect our GatsbyJS with the CMS over GraphQL.
+
+Let's start with installing the plugin:
+<br />
+`composer require markhuot/craftql:^1.0.0`
+<br />
+<br />
+Once it is installed, we need to create an endpoint. Go to your CraftCMS admin and head over to **settings —> CraftQL**.
+Enter the URI for your endpoint and generate a new token. Make sure you check the settings in the token you created. Here you can adjust the power of the token. Should this token be read-only or can you also mutate entries with your token? Since we creating a frontend application, we do not need any mutations here.
```

Comment:
add:
* your api endpoint will be the your CraftCMS URL + URI name (default: api) ... https://your-craftCMS-url.com/api
* do not forget to update the API settings if you add a new section. Otherwise the API does not know that there is a new section (e.g. contact or whatever)